### ISSUE TITLE
Create MTScript 2 parser

### ISSUE THREAD

**Issue by cwisniew** [RPTools/maptool#245]
Create a new parser to handle MTScript 2, this will eventually become the default macro script parser for MapTool while MTScript 1 will be deprecated (but remain for some time for compatibility).

**Comment by mikehenry-se** [comment]
Suggestion for candidate MTScript 2: Jython with Jinja2. 
These two will provide a high-performance and widely used templating engine for MT, enjoying wide support even outside the MT community, and conserve MT development resources for other things. Jython is a Python interpreter written in Java and should be pretty straight-forward to integrate into MT.
True, this is not backward-compatible with current MTScript, but well worth the switch, IMHO, as everything current MTScript can do, can be done in Python with Jinja templates.

**Comment by Azhrei** [comment]
I thought Jython was [essentially a dead tool](https://stackoverflow.com/questions/2351008/when-will-jython-support-python-3)?  It only supports Python 2.7 and last I heard, the devs were not going to be updating it to Python 3.

It would be really cool to be able to use whatever language one wished (Python, Groovy, JavaScript, etc) but the workload for that is pretty high.

**Comment by Phergus** [comment]
It's still Python 2.7 but a recent release has brought it up to Java 11.

**Comment by mikehenry-se** [comment]
I agree that things were looking pretty grim for Jython for a number of years. But with the move to GitHub and the 2.7.2 beta release with slim jars distributed via Maven Central, things are looking up again.
The GitHub move also enables a wider community collaboration even if the current team fizzles.

**Comment by kayila** [comment]
Everything I've ever seen for Jython seems to indicate that it is no longer
a maintained project. I've used it in the past, and loved it, but I would
not want to bring an unmaintained project as a dependency for MapTool.
Especially when it is already using an obsolete, unsupported version of
python.

**Comment by Phergus** [comment]
Again, for accuracy's sake, Jython had two releases this year.   Python 2.7.17 was released 10/19/2019.

**Comment by Azhrei** [comment]
Doesn’t matter if Jython has had releases, the main contributor has said they _want_ a Python 3 build, but they said that 8 years ago.  And Python 2.7 is end of life in June 2020, so it’s essentially dead (as @kayila says, not a good addition at this time).

There are two things that could change this:  if a release came out for Python 3 (and had a few months to settle down), or if MT had a language bindings later that could support any language.

(The ability to write in Python and generate Java `.class` files is cool, but I’ll ultimately not that useful right now.  Maybe later.)

**Comment by mikehenry-se** [comment]
Hmm.. If Python 3 is a requirement then maybe [GraalVM](https://www.graalvm.org/)'s [GraalPython](https://github.com/graalvm/graalpython) could do the trick. Moving to the [GraalVM](https://www.graalvm.org/) also give's a lot of other opportunities as well for MT2 as a whole, such as support for languages besides Python - e.g. Node and Javascript support for dynamic frames and dialogs, etc.

Seems like more work than "merely" dropping in Jython 2.7, but maybe much more worth it in the long run?

**Comment by Azhrei** [comment]
Yes, GraalVM is the current plan.

**Comment by cwisniew** [comment]
Just to be clear though, there is no plan to support any language anyone wants though. It's been discussed many times before and we don't want have to support conversion between arbitrary data types in any language people want to use and what  ever anyone else is using
